src/io/sudhir/switchboard/boards/ImmutableBoard.java

```diff
@@ -37,8 +39,10 @@ static ImmutableBoard create(
   abstract ImmutableBoard parentBoard();
 
   @Override
-  public Stream<Choice> choicesMade() {
-    return historyStream().map(ImmutableBoard::choice).filter(Objects::nonNull);
+  @Memoized
+  public List<Choice> choicesMade() {
+    return historyStream().map(ImmutableBoard::choice).filter(Objects::nonNull).collect(
+        Collectors.toList());
```

Comment:
Make this immutable. Make sure everything that's given to code outside this library is immutable. Since this is a one time operation the perf cost is negligible.

Reply:
Done @sudhirj . However, List is also immutable, is what I think, only ArrayLists are mutable.